### PR TITLE
Introduced tasks and configuration for testing releases.

### PR DESCRIPTION
**Pull request by pomek**
Fixes: #260 

**Requires:** https://github.com/ckeditor/ckeditor5-dev-compiler/pull/23

This PR will introduce new tasks:

* `test:samples:local` - building tests with local modules,
* `test:samples:bundled` - building tests with bundled editor,
* `test:samples` - an alias for above tasks,
* `compile:samples` - building editors (needed for `test:samples:bundled`),
* `compile:samples:clean` - removes built editors. 

For testing it, you need to create some sample files:

**`ckeditor5-typing/docs/samples/typing.js`**
```js
/**
 * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */

/* global console:false, bender: false */

import ClassicEditor from '/ckeditor5/editor-classic/classic.js'; // editor:module

const done = bender.defer();

// sample:start
ClassicEditor.create( document.querySelector( '#editor' ), { // editor:name
	features: [ 'enter', 'typing', 'paragraph', 'undo', 'basic-styles/bold', 'basic-styles/italic' ], // sample:remove|editor:features
	toolbar: [ 'bold', 'italic', 'undo', 'redo' ]
} )
.then( editor => {
	window.editor = editor;
	done(); // sample:remove
}, err => {
	console.error( err.stack );
} );
// sample:end

describe( 'Typing sample', () => {
	it( 'contains default value', () => {
		const output = '<p><em>This</em> is an <strong>editor</strong> instance...</p>';

		expect( editor.getData() ).to.equal( output );
	} );
} );
```

**`ckeditor5-typing/docs/samples/typing.html`**
```html
<div id="editor">
	<p><em>This</em> is an <strong>editor</strong> instance...</p>
</div>
```

**`ckeditor5-basic-samples/docs/samples/basic-styles.js`**
```js
/**
 * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md.
 */

/* global console:false, bender: false */

import ClassicEditor from '/ckeditor5/editor-classic/classic.js'; // editor:module

const done = bender.defer();

// sample:start
ClassicEditor.create( document.querySelector( '#editor' ), { // editor:name
	features: [ 'enter', 'typing', 'paragraph', 'undo', 'basic-styles/bold', 'basic-styles/italic' ], // sample:remove|editor:features
	toolbar: [ 'bold', 'italic', 'undo', 'redo' ]
} )
.then( editor => {
	window.editor = editor;
	done(); // sample:remove
}, err => {
	console.error( err.stack );
} );
// sample:end

describe( 'Basic Styles sample', () => {
	it( 'contains default value', () => {
		const output = '<p><em>This</em> is an <strong>editor</strong> instance...</p>';

		expect( editor.getData() ).to.equal( output );
	} );
} );
```

**`ckeditor5-basic-samples/docs/samples/basic-styles.html`**
```html
<div id="editor">
	<p><em>This</em> is an <strong>editor</strong> instance...</p>
</div>
```

### CODE REVIEW

**Comment by pomek** [comment]
**`r-` because there is still one bug. It will be fixed tomorrow (probably).**

File: gulpfile.js [hunk 1]
```diff
@@ -24,6 +24,15 @@ const config = {
 	// Path to the default configuration file for bundler.
 	BUNDLE_DEFAULT_CONFIG: 'dev/bundles/build-config-standard.js',
 
+	DOCUMENTATION: {
+		// Path to the built editors for samples.
+		EDITORS_DIST: '.docs/assets/scripts/samples',
```
Comment (Reinmar) [inline gulpfile.js]: `.docs` looks like a mistake.

Comment (Reinmar) [inline gulpfile.js]: OK, I can see that you meant to use `.docs`. Then this is bad. You should use some directory inside `build/docs/`. Also, aren't those files needed only temporarily?

Comment (pomek) [inline gulpfile.js]: During this issue, those editors aren't needed. But, in the future we will introduce `docs-builder` and that tool will need to have the editors (for sample pages).

**Comment by Reinmar** [comment]
There are couple of things that I noticed already:

1. The sample tests (not bundled) should be generated together with all other tests (so when using `gulp compile [--watch]`). The goal is to use sample tests as part of integration tests. This, however, may require more coding, so we don't have to work on this in this PR.
2. In the bundled scenario, the editor package is injected directly to test JS file. It should be saved on a disk as a separate file and loaded using `bender-include` directive (like [here](https://github.com/ckeditor/ckeditor-dev/blob/846f2a1a8f45db9b2c59538b59cbbfb1080b84ed/tests/plugins/clipboard/inline.js#L3)).
3. When marking lines (`// sample:remove|editor:features`), I think that comma (followed by a space of course!) would be more suitable than `|` to separate directives.

**Comment by pomek** [comment]
I agree with all the points and I will fix.

**Comment by Reinmar** [comment]
One thing that I want to think through is the naming of the commands. Right now there's lack of consistency and a lack of plan visible, because the tasks were created by many devs at different stages.

**Comment by Reinmar** [comment]
I extracted the first point from https://github.com/ckeditor/ckeditor5/pull/299#issuecomment-240974790 to https://github.com/ckeditor/ckeditor5-dev-compiler/issues/24.

**Comment by Reinmar** [comment]
I pushed one more commit which cleans up few things in the gulpfile. I also reported a few tickets:

* https://github.com/ckeditor/ckeditor5/issues/310
* https://github.com/ckeditor/ckeditor5/issues/309
* https://github.com/ckeditor/ckeditor5/issues/308

**Comment by Reinmar** [comment]
There are couple of things that still need to be taken care of:

* Some garbage is left in `build/dist/` after running `gulp compile:bundled-sample-tests`.
* There's an inconsistency in config options – there's `DOCS_DIR` and `DOCUMENTATION`. Do I understand correctly that the first option will be removed soon? If so, then it can stay there for a while.
* `compiler.compile.tests` should be renamed to `compiler.compile.sampleTests`.

Also, I start to think that we need a new repository for the docs things or something, because the code which bundles editors for release tests don't match any repository that we have now... We should think about this after closing these PRs.

**Comment by pomek** [comment]
@Reinmar,

1. Done.
2. Right, `docs-builder` will unify it.
3. Done.

> I start to think that we need a new repository for the docs things or something

@oskarwrobel [mentioned](https://github.com/ckeditor/ckeditor5-dev-bundler-rollup/pull/15/files/51f8d0d6fce878a6bd016c44581802043282188e#r74061718) about it. I also thought about it.

**Comment by Reinmar** [comment]
OK, I've done some refactoring here and there, moved some code from dependent PRs to this one (temporarily) and merged this code to branch `docs`. I changed deps to use GH links so running `npm i` after cleaning up `node_modules/@ckeditor` should switch all deps correctly.

Let's keep this code out of master until we have the documentation ready.

**Pull request closed without merging**